### PR TITLE
refactor: improve compilation performance

### PR DESCRIPTION
**Pull request by Karrq**
## Motivation
Currently compiling projects can be quite slow, especially for bigger and more complicated projects.
The PR aims to improve the compilation speed for zk contracts substantially.

## Solution
The main change proposed in the PR regards the way we invoke the compiler - where as before we'd invoke the compiler separately for each contract (with its own set of dependencies), we now call the compiler a limited number of times (typically 1) with all the files in the project and the dependencies, drastically reducing compile time and avoiding recompilation of shared dependencies.

Other improvements aim at decreasing memory usage and improving the speed of processing the compiler's output and preparing its input.

## Notes
This PR is still work in progress as the compilation process has been overhauled greatly - there might still be a few bugs and regressions introduced, but before merging these will be resolved

### CODE REVIEW

File: crates/zksync/compiler/src/zksolc/compile.rs [hunk 1]
```diff
@@ -71,16 +75,32 @@ pub type ContractBytecodes = BTreeMap<String, String>;
...
+    base: PathBuf,
 }
 
 impl ZkSolcArtifactPaths {
     pub fn new(filename: PathBuf) -> Self {
-        Self {
-            artifact: filename.join("artifacts.json"),
-            contract_hash: filename.join("contract_hash"),
-        }
+        Self { base: filename }
```
Comment (nbaztec) [inline crates/zksync/compiler/src/zksolc/compile.rs:83]: There's currently a major bug in how we cache and compile artifacts from solc. I'm working on fixing that and aligning the artifact generation (and caching). Can we limit this PR to only the necessary modifications - or perhaps rebase it after the bug is fixed?

Comment (Karrq) [reply crates/zksync/compiler/src/zksolc/compile.rs:83]: We can merge/rebase it after. The process doesn't depend much on the precise way we store artifacts

File: crates/zksync/compiler/src/zksolc/compile.rs [hunk 2]
```diff
@@ -975,90 +1067,80 @@ impl ZkSolc {
...
+        let mut cache = Vec::with_capacity(all_sources.len());
+        all_sources.retain(|path, _| {
+            let relative_path = path.strip_prefix(self.project.root()).unwrap_or(path.as_ref());
+            let is_ignored = self.is_contract_ignored_in_config(relative_path);
+            let is_ignored = is_ignored || additional_filter(path);
+
+            let cached =
+                Self::check_contract_is_cached(artifacts_paths, path).ok().and_then(|r| r.0);
+
+            // prune ignored or cached contractacs
```
Comment (Jrigada) [inline crates/zksync/compiler/src/zksolc/compile.rs:1105]:
```suggestion
            // prune ignored or cached contracts
```

**Review by Jrigada — APPROVED** [review]

**Review by HermanObst — CHANGES_REQUESTED** [review]
**Issue**: When running a repo that have missing-libraries, the compilation pass but without detecting them. 
To reproduce, try to compile `aave-v3-origin` from scratch and the compilation will finish without noticing the missing libraries.

This need to be investigated further. Here are my compilation logs;
```
herman@srv627971:~/aave-v3-origin$ RUST_BACKTRACE=1 RUST_LOG=debug ../foundry-zksync/target/debug/forge build --zksync
[⠃] Compiling...2024-05-15T20:27:03.472569Z DEBUG compile_with: foundry_common::compile: compiling project
2024-05-15T20:27:03.475661Z DEBUG compile_with:compile:version: foundry_compilers::compile: getting Solc version cmd="/home/herman/.svm/0.8.19/solc-0.8.19" "--version"
2024-05-15T20:27:03.477341Z DEBUG compile_with:compile:version: foundry_compilers::compile: version=0.8.19+commit.7dd6d404.Linux.gcc
[⠊] Compiling...2024-05-15T20:27:03.615557Z DEBUG compile_with:compile:compile: foundry_compilers::compile: compiling cmd=cd "/home/herman/aave-v3-origin" && "/home/herman/.svm/0.8.19/solc-0.8.19" "--base-path" "/home/herman/aave-v3-origin" "--allow-paths" "/home/herman/aave-v3-origin,/home/herman/aave-v3-origin/lib" "--standard-json"
2024-05-15T20:27:03.615735Z DEBUG compile_with:compile:compile: foundry_compilers::compile: spawned
2024-05-15T20:27:03.639980Z DEBUG compile_with:compile:compile: foundry_compilers::compile: wrote JSON input to stdin
[⠒] Compiling...
[⠒] Compiling 336 files with 0.8.192024-05-15T20:27:52.653171Z DEBUG compile_with:compile:compile: foundry_compilers::compile: finished output.status=exit status: 0 output.stderr=""
[⠒] Compiling 336 files with 0.8.19
[⠢] Solc 0.8.19 finished in 49.95s2024-05-15T20:27:54.680441Z DEBUG compile_with: foundry_common::compile: finished compiling in 51.208s
[⠆] Solc 0.8.19 finished in 49.95s
Compiler run successful with warnings:
Warning (5740): Unreachable code.
  --> src/periphery/contracts/dependencies/openzeppelin/ReentrancyGuard.sol:61:5:
   |
61 |     _status = _NOT_ENTERED;
   |     ^^^^^^^^^^^^^^^^^^^^^^

2024-05-15T20:27:54.827165Z DEBUG version: foundry_compilers::compile: getting Solc version cmd="/home/herman/.svm/0.8.23/solc-0.8.23" "--version"
2024-05-15T20:27:54.828605Z DEBUG version: foundry_compilers::compile: version=0.8.23+commit.f704f362.Linux.gcc
2024-05-15T20:27:54.838236Z DEBUG version: foundry_compilers::compile: getting Solc version cmd="/home/herman/.svm/0.8.23/solc-0.8.23" "--version"
2024-05-15T20:27:54.838983Z DEBUG version: foundry_compilers::compile: version=0.8.23+commit.f704f362.Linux.gcc
2024-05-15T20:27:54.853354Z  INFO foundry_zksync_compiler::zksolc::compile: 
Compiling 336 files... solc="/home/herman/.svm/0.8.23/solc-0.8.23"
⠕ Compiling 336 files..
```

Also, after the compilation we are not outputting `zkout` directory

cc/ @Karrq @nbaztec

**Comment by HermanObst** [comment]
Lets wait to Monday sync before merging this :)

**Comment by Karrq** [comment]
This PR is still valid and will receive updates and patches, at least until #368 is merged, but will otherwise not be merged with dev

**Pull request closed without merging**